notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
+// Show shows the notification.
+func (n *Notification) Show() {
+	if lib == nil {
+		log.Error("notify: library not properly loaded")
 		return
 	}
 
-	notificationsLock.Lock()
-	var n *Notification
-	for _, elem := range notifications {
-		if elem.GUID == data.id {
-			n = elem
-			break
-		}
-	}
-	notificationsLock.Unlock()
-	if n == nil {
-		return
+	if !libInitialized {
+		log.Error("notify: not initialized")
 	}
+
+	// Lock notification
 	n.Lock()
 	defer n.Unlock()
 
-	switch data.action {
-	case "dismissed": // do nothing
-	case "clicked": // like dismissed: do nothing
-	case "buttonClicked":
-		for _, actions := range n.AvailableActions {
-			if actions.Text == data.button {
-				n.SelectAction(actions.ID)
-				return
-			}
+	// Lock library
+	lib.Lock()
+	defer lib.Unlock()
+
+	title, _ := windows.UTF16PtrFromString(n.Title)
+	message, _ := windows.UTF16PtrFromString(n.Message)
+	titleP := unsafe.Pointer(title)
+	messageP := unsafe.Pointer(message)
+	if lib != nil {
+		// Create new notification object
+		notificationTemplate, _, err := lib.createNotification.Call(uintptr(titleP), uintptr(messageP))
+		if notificationTemplate == 0 {
+			log.Errorf("notify: failed to create notification: %q", err)
 		}
+		// Make sure memory is freed when done
+		defer func() { _, _, _ = lib.deleteNotification.Call(notificationTemplate) }()
 
-		log.Warningf("failed to handle button click: button %s is reported to have been clicked but is no registered button. Available options: %+v", data.button, n.AvailableActions)
-	default:
-		log.Debugf("failed to handle SnoreToast-Action %s: not implemented yet", data.action)
-	}
-}
-
-// Snoretoast helper-Functions:
-
-// Verifies that the Argument contians no Semicolon which would be difficult (in some cases impossible) to parse in a Pipe-Response
-func verifySnoreToastArgumentSyntax(arg string) error {
-	if strings.Contains(arg, ";") {
-		return fmt.Errorf("the SnoreToast-Argument %s would contain a semicolon which would screw up the pipe responses", arg)
-	}
-	return nil
-}
-
-func parseSnoreToastActionCallback(in string) actionCallback {
-	ret := actionCallback{}
-
-	for _, elem := range strings.Split(in, ";") {
-		if elem == "" {
-			continue
+		// Set Portmaster icon.
+		iconLocation, err := ensureAppIcon()
+		if err != nil {
+			log.Warningf("notify: failed to write icon: %s", err)
 		}
-
-		elemSplit := strings.SplitN(elem, "=", 2)
-		if len(elemSplit) != 2 {
-			log.Warningf("failing to parse snoretoast-Response %s into Key=Value-Pair; Response: %s", elem, in)
-			continue
+		iconPathUTF, _ := windows.UTF16PtrFromString(iconLocation)
+		iconPathP := unsafe.Pointer(iconPathUTF)
+		_, _, _ = lib.setImage.Call(notificationTemplate, uintptr(iconPathP))
+
+		// Set all the required actions
+		for _, action := range n.AvailableActions {
+			textUTF, _ := windows.UTF16PtrFromString(action.Text)
+			textP := unsafe.Pointer(textUTF)
+			_, _, _ = lib.addButton.Call(notificationTemplate, uintptr(textP))
```

Comment:
Should we at least warn if there is an error?

Reply:
the only time that it will fail is when the parameters are not valid, which should not happen in this case.
I dont think it is necessary in this case. It will just add more clutter to the code.